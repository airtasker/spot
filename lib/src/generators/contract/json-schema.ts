import assertNever from "assert-never";
import * as YAML from "js-yaml";
import { Api, normalizedObjectType, Type, Types } from "../../models";
import compact = require("lodash/compact");

export function generateJsonSchema(api: Api, format: "json" | "yaml") {
  const contract = jsonSchema(api);
  switch (format) {
    case "json":
      return JSON.stringify(contract, null, 2);
    case "yaml":
      return YAML.safeDump(contract);
    default:
      throw assertNever(format);
  }
}

export function jsonSchema(api: Api): JsonSchema {
  return {
    $schema: "http://json-schema.org/draft-07/schema#",
    definitions: Object.entries(api.types).reduce(
      (acc, [typeName, type]) => {
        acc[typeName] = jsonTypeSchema(api.types, type);
        return acc;
      },
      {} as { [typeName: string]: JsonSchemaType }
    )
  };
}

export function jsonTypeSchema(types: Types, type: Type): JsonSchemaType {
  switch (type.kind) {
    case "void":
    case "null":
      return {
        type: "null"
      };
    case "boolean":
      return {
        type: "boolean"
      };
    case "boolean-constant":
      return {
        type: "boolean",
        const: type.value
      };
    case "date":
    case "date-time":
    case "string":
      return {
        type: "string"
      };
    case "string-constant":
      return {
        type: "string",
        const: type.value
      };
    case "number":
    case "int32":
    case "int64":
    case "float":
    case "double":
      return {
        type: "number"
      };
    case "integer-constant":
      return {
        type: "integer",
        const: type.value
      };
    case "object":
      return Object.entries(normalizedObjectType(types, type)).reduce(
        (acc, [name, type]) => {
          if (type.kind === "optional") {
            type = type.optional;
          } else {
            acc.required.push(name);
          }
          const schemaType = jsonTypeSchema(types, type);
          if (schemaType) {
            acc.properties[name] = schemaType;
          }
          return acc;
        },
        {
          type: "object",
          properties: {},
          required: []
        } as JsonSchemaObject & { required: string[] }
      );
    case "array":
      const itemsType = jsonTypeSchema(types, type.elements);
      if (!itemsType) {
        throw new Error(`Unsupported void array`);
      }
      return {
        type: "array",
        items: itemsType
      };
    case "optional":
      throw new Error(`Unsupported top-level optional type`);
    case "union":
      const unionTypes = type.types.map(t => jsonTypeSchema(types, t));
      const withoutNullTypes = compact(unionTypes);
      if (withoutNullTypes.length !== unionTypes.length) {
        throw new Error(`Unsupported void type in union`);
      }
      return {
        oneOf: withoutNullTypes
      };
    case "type-reference":
      return {
        $ref: `#/definitions/${type.typeName}`
      };
    default:
      throw assertNever(type);
  }
}

export interface JsonSchema {
  $schema: "http://json-schema.org/draft-07/schema#";
  definitions: {
    [typeName: string]: JsonSchemaType;
  };
}

export type JsonSchemaType =
  | JsonSchemaObject
  | JsonSchemaArray
  | JsonSchemaOneOf
  | JsonSchemaNull
  | JsonSchemaString
  | JsonSchemaNumber
  | JsonSchemaInteger
  | JsonSchemaBoolean
  | JsonSchemaTypeReference;

export interface JsonSchemaObject {
  type: "object";
  properties: {
    [name: string]: JsonSchemaType;
  };
  required?: string[];
}

export interface JsonSchemaArray {
  type: "array";
  items: JsonSchemaType;
}

export interface JsonSchemaOneOf {
  oneOf: JsonSchemaType[];
  discriminator?: {
    propertyName: string;
    mapping: {
      [value: string]: JsonSchemaType;
    };
  };
}

export interface JsonSchemaNull {
  type: "null";
}

export interface JsonSchemaString {
  type: "string";
  const?: string;
  enum?: string[];
}

export interface JsonSchemaNumber {
  type: "number";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaInteger {
  type: "integer";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaBoolean {
  type: "boolean";
  const?: boolean;
}

export interface JsonSchemaTypeReference {
  $ref: string;
}
