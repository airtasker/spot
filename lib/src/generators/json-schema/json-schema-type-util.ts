import assertNever from "assert-never";
import {
  areBooleanLiteralTypes,
  areFloatLiteralTypes,
  areIntLiteralTypes,
  areStringLiteralTypes,
  BooleanLiteralType,
  FloatLiteralType,
  IntLiteralType,
  isBooleanLiteralType,
  isFloatLiteralType,
  isIntLiteralType,
  isNotLiteralType,
  isStringLiteralType,
  StringLiteralType,
  Type,
  TypeKind
} from "../../types";
import { JsonSchemaObject, JsonSchemaType } from "./json-schema-specification";

/**
 * Generate a JSON Schema type definition. `objectAdditionalProperties` may
 * be used to configure whether additional properties should be allowed on
 * object types. This can be useful for data validation purposes where
 * property validation can be strict or lenient.
 *
 * @param type a contract type
 * @param objectAdditionalProperties whether to allow additional properties for objects
 */
export function typeToJsonSchemaType(
  type: Type,
  objectAdditionalProperties: boolean = true // TODO: expose proper configuration object
): JsonSchemaType {
  switch (type.kind) {
    case TypeKind.NULL:
      return {
        type: "null"
      };
    case TypeKind.BOOLEAN:
      return {
        type: "boolean"
      };
    case TypeKind.BOOLEAN_LITERAL:
      return {
        type: "boolean",
        const: type.value
      };
    case TypeKind.DATE:
      return { type: "string", format: "date" };
    case TypeKind.DATE_TIME:
      return { type: "string", format: "date-time" };
    case TypeKind.STRING:
      return {
        type: "string"
      };
    case TypeKind.STRING_LITERAL:
      return {
        type: "string",
        const: type.value
      };
    case TypeKind.FLOAT:
    case TypeKind.DOUBLE:
      return {
        type: "number"
      };
    case TypeKind.FLOAT_LITERAL:
      return {
        type: "number",
        const: type.value
      };
    case TypeKind.INT32:
    case TypeKind.INT64:
      return {
        type: "integer"
      };
    case TypeKind.INT_LITERAL:
      return {
        type: "integer",
        const: type.value
      };
    case TypeKind.OBJECT:
      return type.properties.reduce<JsonSchemaObject & { required: string[] }>(
        (acc, property) => {
          if (!property.optional) {
            acc.required.push(property.name);
          }
          acc.properties[property.name] = typeToJsonSchemaType(
            property.type,
            objectAdditionalProperties
          );
          return acc;
        },
        {
          type: "object",
          properties: {},
          required: [],
          additionalProperties: objectAdditionalProperties
        }
      );
    case TypeKind.ARRAY:
      return {
        type: "array",
        items: typeToJsonSchemaType(
          type.elementType,
          objectAdditionalProperties
        )
      };
    case TypeKind.UNION:
      const elements = type.types;
      if (elements.length === 0) throw new Error("Union type has no elements");
      if (elements.length === 1) return typeToJsonSchemaType(elements[0]);

      if (
        areBooleanLiteralTypes(elements) ||
        areStringLiteralTypes(elements) ||
        areFloatLiteralTypes(elements) ||
        areIntLiteralTypes(elements)
      ) {
        return singleTypeLiteralsToSchema(elements);
      } else {
        // Guaranteed oneOf

        const oneOfElements = elements
          .filter(isNotLiteralType)
          .map(t => typeToJsonSchemaType(t, objectAdditionalProperties));

        const booleanLiterals = elements.filter(isBooleanLiteralType);
        if (booleanLiterals.length > 0) {
          oneOfElements.push(singleTypeLiteralsToSchema(booleanLiterals));
        }

        const stringLiterals = elements.filter(isStringLiteralType);
        if (stringLiterals.length > 0) {
          oneOfElements.push(singleTypeLiteralsToSchema(stringLiterals));
        }

        const floatLiterals = elements.filter(isFloatLiteralType);
        if (floatLiterals.length > 0) {
          oneOfElements.push(singleTypeLiteralsToSchema(floatLiterals));
        }

        const integerLiterals = elements.filter(isIntLiteralType);
        if (integerLiterals.length > 0) {
          oneOfElements.push(singleTypeLiteralsToSchema(integerLiterals));
        }

        return {
          oneOf: oneOfElements
        };
      }
    case TypeKind.REFERENCE:
      return {
        $ref: `#/definitions/${type.name}`
      };
    default:
      throw assertNever(type);
  }
}

function singleTypeLiteralsToSchema(
  literals:
    | BooleanLiteralType[]
    | StringLiteralType[]
    | FloatLiteralType[]
    | IntLiteralType[]
): JsonSchemaType {
  switch (literals.length) {
    case 0:
      throw new Error("no literals found");
    case 1:
      return typeToJsonSchemaType(literals[0]);
    default:
      if (areBooleanLiteralTypes(literals)) {
        literals.map(e => e.value);
        return {
          type: "boolean",
          enum: literals.map(e => e.value)
        };
      } else if (areStringLiteralTypes(literals)) {
        return {
          type: "string",
          enum: literals.map(e => e.value)
        };
      } else if (areFloatLiteralTypes(literals)) {
        return {
          type: "number",
          enum: literals.map(e => e.value)
        };
      } else if (areIntLiteralTypes(literals)) {
        return {
          type: "integer",
          enum: literals.map(e => e.value)
        };
      } else {
        throw new Error("Unknown literals");
      }
  }
}
