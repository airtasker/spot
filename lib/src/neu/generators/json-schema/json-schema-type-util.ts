import assertNever from "assert-never";
import {
  areBooleanLiteralTypes,
  areFloatLiteralTypes,
  areIntLiteralTypes,
  areStringLiteralTypes,
  isBooleanLiteralType,
  isFloatLiteralType,
  isIntLiteralType,
  isNotLiteralType,
  isStringLiteralType,
  Type,
  TypeKind
} from "../../types";
import { JsonSchemaObject, JsonSchemaType } from "./json-schema-specification";

/**
 * Generate a JSON Schema type definition. `objectAdditionalProperties` may
 * be used to configure whether additional properties should be allowed on
 * object types. This can be useful for data validation purposes where
 * property validation can be strict or lenient.
 *
 * @param type a contract type
 * @param objectAdditionalProperties whether to allow additional properties for objects
 */
export function typeToJsonSchemaType(
  type: Type,
  objectAdditionalProperties: boolean = true // TODO: expose proper configuration object
): JsonSchemaType {
  switch (type.kind) {
    case TypeKind.NULL:
      return {
        type: "null"
      };
    case TypeKind.BOOLEAN:
      return {
        type: "boolean"
      };
    case TypeKind.BOOLEAN_LITERAL:
      return {
        type: "boolean",
        const: type.value
      };
    case TypeKind.DATE:
      return { type: "string", format: "date" };
    case TypeKind.DATE_TIME:
      return { type: "string", format: "date-time" };
    case TypeKind.STRING:
      return {
        type: "string"
      };
    case TypeKind.STRING_LITERAL:
      return {
        type: "string",
        const: type.value
      };
    case TypeKind.FLOAT:
    case TypeKind.DOUBLE:
      return {
        type: "number"
      };
    case TypeKind.FLOAT_LITERAL:
      return {
        type: "number",
        const: type.value
      };
    case TypeKind.INT32:
    case TypeKind.INT64:
      return {
        type: "integer"
      };
    case TypeKind.INT_LITERAL:
      return {
        type: "integer",
        const: type.value
      };
    case TypeKind.OBJECT:
      return type.properties.reduce<JsonSchemaObject & { required: string[] }>(
        (acc, property) => {
          if (!property.optional) {
            acc.required.push(property.name);
          }
          acc.properties[property.name] = typeToJsonSchemaType(
            property.type,
            objectAdditionalProperties
          );
          return acc;
        },
        {
          type: "object",
          properties: {},
          required: [],
          additionalProperties: objectAdditionalProperties
        }
      );
    case TypeKind.ARRAY:
      return {
        type: "array",
        items: typeToJsonSchemaType(
          type.elementType,
          objectAdditionalProperties
        )
      };
    case TypeKind.UNION:
      const elements = type.types;
      if (areBooleanLiteralTypes(elements)) {
        return {
          type: "boolean",
          enum: elements.map(e => e.value)
        };
      } else if (areStringLiteralTypes(elements)) {
        return {
          type: "string",
          enum: elements.map(e => e.value)
        };
      } else if (areFloatLiteralTypes(elements)) {
        return {
          type: "number",
          enum: elements.map(e => e.value)
        };
      } else if (areIntLiteralTypes(elements)) {
        return {
          type: "integer",
          enum: elements.map(e => e.value)
        };
      } else {
        // Guaranteed oneOf

        const oneOfElements = elements
          .filter(isNotLiteralType)
          .map(t => typeToJsonSchemaType(t, objectAdditionalProperties));

        const booleanLiterals = elements.filter(isBooleanLiteralType);
        switch (booleanLiterals.length) {
          case 0:
            break;
          case 1:
            oneOfElements.push(typeToJsonSchemaType(booleanLiterals[0]));
            break;
          default:
            oneOfElements.push({
              type: "boolean",
              enum: booleanLiterals.map(b => b.value)
            });
        }

        const stringLiterals = elements.filter(isStringLiteralType);
        switch (stringLiterals.length) {
          case 0:
            break;
          case 1:
            oneOfElements.push(typeToJsonSchemaType(stringLiterals[0]));
            break;
          default:
            oneOfElements.push({
              type: "string",
              enum: stringLiterals.map(s => s.value)
            });
        }

        const floatLiterals = elements.filter(isFloatLiteralType);
        switch (floatLiterals.length) {
          case 0:
            break;
          case 1:
            oneOfElements.push(typeToJsonSchemaType(floatLiterals[0]));
            break;
          default:
            oneOfElements.push({
              type: "number",
              enum: floatLiterals.map(f => f.value)
            });
        }

        const integerLiterals = elements.filter(isIntLiteralType);
        switch (integerLiterals.length) {
          case 0:
            break;
          case 1:
            oneOfElements.push(typeToJsonSchemaType(integerLiterals[0]));
            break;
          default:
            oneOfElements.push({
              type: "integer",
              enum: integerLiterals.map(i => i.value)
            });
        }

        return {
          oneOf: oneOfElements
        };
      }
    case TypeKind.REFERENCE:
      return {
        $ref: `#/definitions/${type.name}`
      };
    default:
      throw assertNever(type);
  }
}
