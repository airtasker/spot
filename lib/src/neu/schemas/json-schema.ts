export interface JsonSchema {
  $schema: "http://json-schema.org/draft-07/schema#";
  definitions: {
    [typeName: string]: JsonSchemaType;
  };
}

export type JsonSchemaType =
  | JsonSchemaObject
  | JsonSchemaArray
  | JsonSchemaOneOf
  | JsonSchemaNull
  | JsonSchemaString
  | JsonSchemaNumber
  | JsonSchemaInteger
  | JsonSchemaBoolean
  | JsonSchemaTypeReference;

export interface JsonSchemaObject {
  type: "object";
  properties: {
    [name: string]: JsonSchemaType;
  };
  required?: string[];
  additionalProperties: boolean;
}

export interface JsonSchemaArray {
  type: "array";
  items: JsonSchemaType;
}

export interface JsonSchemaOneOf {
  oneOf: JsonSchemaType[];
  discriminator?: {
    propertyName: string;
    mapping: {
      [value: string]: JsonSchemaType;
    };
  };
}

export interface JsonSchemaNull {
  type: "null";
}

export interface JsonSchemaString {
  type: "string";
  format?: string;
  const?: string;
  enum?: string[];
}

export interface JsonSchemaNumber {
  type: "number";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaInteger {
  type: "integer";
  const?: number;
  enum?: number[];
}

export interface JsonSchemaBoolean {
  type: "boolean";
  const?: boolean;
}

export interface JsonSchemaTypeReference {
  $ref: string;
}
