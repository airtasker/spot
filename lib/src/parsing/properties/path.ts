import ts from "typescript";
import { DynamicPathComponent, PathComponent, VOID } from "../../models";
import { extractSingleDecorator } from "../decorators";
import {
  isObjectLiteral,
  isStringLiteral,
  ObjectLiteral
} from "../literal-parser";
import { panic } from "../panic";
import { extractParameterType } from "./parameter-type";

/**
 * Returns the path of an endpoint.
 *
 * For example:
 * ```
 * @endpoint({
 *   method: "GET",
 *   path: "/users/:userId"
 * })
 * getUser(@pathParam() userId: string): User {
 *   ...
 * }
 * ```
 *
 * will return a list of path components with:
 * - a static component ("/users/")
 * - a dynamic component named userId of type string
 */
export function extractPath(
  sourceFile: ts.SourceFile,
  methodDeclaration: ts.MethodDeclaration,
  endpointDescription: ObjectLiteral
): PathComponent[] {
  const pathLiteral = endpointDescription.properties["path"];
  if (!isStringLiteral(pathLiteral)) {
    throw panic(
      `Invalid path in endpoint description: ${methodDeclaration.getText(
        sourceFile
      )}`
    );
  }
  const path = pathLiteral.text;
  const pathComponents: PathComponent[] = [];
  const dynamicPathComponents: { [name: string]: DynamicPathComponent } = {};
  if (path.length > 0) {
    let componentStartPosition = 0;
    do {
      if (path.charAt(componentStartPosition) === ":") {
        // The parameter name extends until a character that isn't a valid name.
        const nextNonNamePositionRelative = path
          .substr(componentStartPosition + 1)
          .search(/[^a-z0-9_]/gi);
        const dynamicPathComponent: DynamicPathComponent = {
          kind: "dynamic",
          name: path.substr(
            componentStartPosition + 1,
            nextNonNamePositionRelative === -1
              ? undefined
              : nextNonNamePositionRelative
          ),
          type: VOID,
          description: ""
        };
        pathComponents.push(dynamicPathComponent);
        dynamicPathComponents[dynamicPathComponent.name] = dynamicPathComponent;
        componentStartPosition =
          nextNonNamePositionRelative === -1
            ? -1
            : componentStartPosition + 1 + nextNonNamePositionRelative;
      } else {
        // The static component extends until the next parameter, which starts with ":".
        const nextColumnPosition = path.indexOf(":", componentStartPosition);
        pathComponents.push({
          kind: "static",
          content: path.substring(
            componentStartPosition,
            nextColumnPosition === -1 ? undefined : nextColumnPosition
          )
        });
        componentStartPosition = nextColumnPosition;
      }
    } while (componentStartPosition !== -1);
  }
  for (const parameter of methodDeclaration.parameters) {
    processPathParameter(
      sourceFile,
      methodDeclaration,
      parameter,
      dynamicPathComponents
    );
  }
  return pathComponents;
}

function processPathParameter(
  sourceFile: ts.SourceFile,
  methodDeclaration: ts.MethodDeclaration,
  parameter: ts.ParameterDeclaration,
  dynamicPathComponents: { [name: string]: DynamicPathComponent }
) {
  const pathParamDecorator = extractSingleDecorator(
    sourceFile,
    parameter,
    "pathParam"
  );
  if (!pathParamDecorator) {
    return;
  }
  const name = parameter.name.getText(sourceFile);
  if (dynamicPathComponents[name]) {
    dynamicPathComponents[name].type = extractParameterType(
      sourceFile,
      parameter
    );
  } else {
    throw panic(
      `Found a path parameter that isn't present in path. Expected one of [${Object.keys(
        dynamicPathComponents
      ).join(", ")}], got this instead: ${name}`
    );
  }
  if (pathParamDecorator.arguments.length > 1) {
    throw panic(
      `Expected exactly one or no arguments for @pathParam(), got ${
        pathParamDecorator.arguments.length
      }`
    );
  }
  if (pathParamDecorator.arguments.length === 1) {
    const pathParamDescription = pathParamDecorator.arguments[0];
    if (!isObjectLiteral(pathParamDescription)) {
      throw panic(
        `@pathParam() expects an object literal, got this instead: ${methodDeclaration.getText(
          sourceFile
        )}`
      );
    }
    const descriptionProperty = pathParamDescription.properties["description"];
    if (!descriptionProperty || !isStringLiteral(descriptionProperty)) {
      throw panic(
        `@pathParam() expects a string description, got this instead: ${methodDeclaration.getText(
          sourceFile
        )}`
      );
    }
    dynamicPathComponents[name].description = descriptionProperty.text;
  }
}
