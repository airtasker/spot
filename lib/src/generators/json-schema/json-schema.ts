import { Contract } from "../../definitions";
import { JsonSchema, JsonSchemaType } from "./json-schema-specification";
import { typeToJsonSchemaType } from "./json-schema-type-util";

export function generateJsonSchema(contract: Contract): JsonSchema {
  return {
    $schema: "http://json-schema.org/draft-07/schema#",
    definitions: contract.types.reduce<{
      [typeName: string]: JsonSchemaType;
    }>((acc, typeNode) => {
      acc[typeNode.name] = typeToJsonSchemaType(typeNode.typeDef.type);
      return acc;
    }, {})
  };
}
