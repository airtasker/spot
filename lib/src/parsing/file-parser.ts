import * as fs from "fs-extra";
import * as path from "path";
import * as ts from "typescript";
import { Api } from "../models";
import { validate } from "../validator";
import { parseApiClass } from "./api-class-parser";
import { extractSingleDecorator } from "./decorators";
import { panic } from "./panic";
import { extractObjectType, extractType } from "./type-parser";

/**
 * Parses a TypeScript source file, as well as any other TypeScript files it imports recursively.
 *
 * This should be smart enough not to get stuck into an infinite loop for circular dependencies.
 */
export async function parsePath(sourcePath: string): Promise<Api> {
  const api: Api = {
    endpoints: {},
    types: {}
  };
  await parseFileRecursively(api, new Set(), sourcePath);
  const errors = validate(api);
  if (errors.length > 0) {
    throw panic(errors.join("\n"));
  }
  return api;
}

async function parseFileRecursively(
  api: Api,
  visitedPaths: Set<string>,
  sourcePath: string
): Promise<void> {
  if (!(await fs.existsSync(sourcePath))) {
    if (await fs.existsSync(sourcePath + ".ts")) {
      sourcePath += ".ts";
    } else {
      throw panic(`No source file found at ${sourcePath}`);
    }
  }
  if (
    visitedPaths.has(sourcePath) ||
    path.resolve(sourcePath).startsWith(__dirname)
  ) {
    return;
  } else {
    visitedPaths.add(sourcePath);
  }
  const fileContent = await fs.readFile(sourcePath, "utf8");
  const sourceFile = ts.createSourceFile(
    path.basename(sourcePath),
    fileContent,
    ts.ScriptTarget.Latest
  );
  for (const statement of sourceFile.statements) {
    if (ts.isImportDeclaration(statement)) {
      if (!ts.isStringLiteral(statement.moduleSpecifier)) {
        throw panic(
          `Unsupported import statement: ${statement.moduleSpecifier.getText(
            sourceFile
          )}`
        );
      }
      const importPath = statement.moduleSpecifier.text;
      if (!importPath.startsWith(".")) {
        // This is not a relative import, we'll ignore it.
        continue;
      }
      await parseFileRecursively(
        api,
        visitedPaths,
        path.join(sourcePath, "..", importPath)
      );
    } else if (ts.isClassDeclaration(statement)) {
      const apiDecorator = extractSingleDecorator(sourceFile, statement, "api");
      if (apiDecorator) {
        parseApiClass(sourceFile, statement, api);
      }
    } else if (ts.isTypeAliasDeclaration(statement)) {
      const name = statement.name.getText(sourceFile);
      api.types[name] = extractType(sourceFile, statement.type);
    } else if (ts.isInterfaceDeclaration(statement)) {
      const name = statement.name.getText(sourceFile);
      api.types[name] = extractObjectType(sourceFile, statement);
    }
  }
}
