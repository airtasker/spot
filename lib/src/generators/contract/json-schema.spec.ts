import * as fs from "fs-extra";
import * as path from "path";
import {
  arrayType,
  BOOLEAN,
  booleanConstant,
  DOUBLE,
  FLOAT,
  INT32,
  INT64,
  integerConstant,
  NULL,
  NUMBER,
  objectType,
  optionalType,
  STRING,
  stringConstant,
  typeReference,
  unionType,
  VOID
} from "../../models";
import { parsePath } from "../../parsing/file-parser";
import { generateJsonSchema, jsonTypeSchema } from "./json-schema";

const EXAMPLES_DIR = path.join(
  __dirname,
  "..",
  "..",
  "..",
  "..",
  "examples",
  "src"
);

describe("JSON Schema generator", () => {
  describe("produces valid code", () => {
    for (const testCaseName of fs.readdirSync(EXAMPLES_DIR)) {
      if (!fs.lstatSync(path.join(EXAMPLES_DIR, testCaseName)).isDirectory()) {
        continue;
      }
      test(testCaseName, async () => {
        const api = await parsePath(
          path.join(EXAMPLES_DIR, testCaseName, `${testCaseName}-api.ts`)
        );
        expect(generateJsonSchema(api, "json")).toMatchSnapshot("json");
        expect(generateJsonSchema(api, "yaml")).toMatchSnapshot("yaml");
      });
    }
  });

  describe("generates type validator", () => {
    test("void", () => {
      expect(jsonTypeSchema(VOID)).toMatchInlineSnapshot(`
Object {
  "type": "null",
}
`);
    });

    test("null", () => {
      expect(jsonTypeSchema(NULL)).toMatchInlineSnapshot(`
Object {
  "type": "null",
}
`);
    });

    test("boolean", () => {
      expect(jsonTypeSchema(BOOLEAN)).toMatchInlineSnapshot(`
Object {
  "type": "boolean",
}
`);
    });

    test("boolean constant", () => {
      expect(jsonTypeSchema(booleanConstant(true))).toMatchInlineSnapshot(`
Object {
  "const": true,
  "type": "boolean",
}
`);
      expect(jsonTypeSchema(booleanConstant(false))).toMatchInlineSnapshot(`
Object {
  "const": false,
  "type": "boolean",
}
`);
    });

    test("string", () => {
      expect(jsonTypeSchema(STRING)).toMatchInlineSnapshot(`
Object {
  "type": "string",
}
`);
    });

    test("string constant", () => {
      expect(jsonTypeSchema(stringConstant("some constant")))
        .toMatchInlineSnapshot(`
Object {
  "const": "some constant",
  "type": "string",
}
`);
    });

    test("number", () => {
      expect(jsonTypeSchema(NUMBER)).toMatchInlineSnapshot(`
Object {
  "type": "number",
}
`);
    });

    test("int32", () => {
      expect(jsonTypeSchema(INT32)).toMatchInlineSnapshot(`
Object {
  "type": "number",
}
`);
    });

    test("int64", () => {
      expect(jsonTypeSchema(INT64)).toMatchInlineSnapshot(`
Object {
  "type": "number",
}
`);
    });

    test("float", () => {
      expect(jsonTypeSchema(FLOAT)).toMatchInlineSnapshot(`
Object {
  "type": "number",
}
`);
    });

    test("double", () => {
      expect(jsonTypeSchema(DOUBLE)).toMatchInlineSnapshot(`
Object {
  "type": "number",
}
`);
    });

    test("integer constant", () => {
      expect(jsonTypeSchema(integerConstant(0))).toMatchInlineSnapshot(`
Object {
  "const": 0,
  "type": "integer",
}
`);
      expect(jsonTypeSchema(integerConstant(123))).toMatchInlineSnapshot(`
Object {
  "const": 123,
  "type": "integer",
}
`);
      expect(jsonTypeSchema(integerConstant(-1000))).toMatchInlineSnapshot(`
Object {
  "const": -1000,
  "type": "integer",
}
`);
    });

    test("object", () => {
      expect(jsonTypeSchema(objectType({}))).toMatchInlineSnapshot(`
Object {
  "properties": Object {},
  "required": Array [],
  "type": "object",
}
`);
      expect(
        jsonTypeSchema(
          objectType({
            singleField: NUMBER
          })
        )
      ).toMatchInlineSnapshot(`
Object {
  "properties": Object {
    "singleField": Object {
      "type": "number",
    },
  },
  "required": Array [
    "singleField",
  ],
  "type": "object",
}
`);
      expect(
        jsonTypeSchema(
          objectType({
            field1: NUMBER,
            field2: STRING,
            field3: optionalType(BOOLEAN)
          })
        )
      ).toMatchInlineSnapshot(`
Object {
  "properties": Object {
    "field1": Object {
      "type": "number",
    },
    "field2": Object {
      "type": "string",
    },
    "field3": Object {
      "type": "boolean",
    },
  },
  "required": Array [
    "field1",
    "field2",
  ],
  "type": "object",
}
`);
    });

    test("array", () => {
      expect(jsonTypeSchema(arrayType(STRING))).toMatchInlineSnapshot(`
Object {
  "items": Object {
    "type": "string",
  },
  "type": "array",
}
`);
    });

    test("optional", () => {
      expect(() => jsonTypeSchema(optionalType(STRING))).toThrowError(
        "Unsupported top-level optional type"
      );
    });

    test("union", () => {
      expect(jsonTypeSchema(unionType(STRING, NUMBER, BOOLEAN)))
        .toMatchInlineSnapshot(`
Object {
  "oneOf": Array [
    Object {
      "type": "string",
    },
    Object {
      "type": "number",
    },
    Object {
      "type": "boolean",
    },
  ],
}
`);
    });

    test("type reference", () => {
      expect(jsonTypeSchema(typeReference("OtherType"))).toMatchInlineSnapshot(`
Object {
  "$ref": "#/definitions/OtherType",
}
`);
    });
  });
});
