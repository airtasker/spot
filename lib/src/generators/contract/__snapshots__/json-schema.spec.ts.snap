// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSON Schema generator produces valid code multi-file: json 1`] = `
"{
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"definitions\\": {
    \\"CreateUserRequest\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        }
      },
      \\"required\\": [
        \\"name\\"
      ]
    },
    \\"CreateUserResponse\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"success\\": {
          \\"type\\": \\"boolean\\"
        }
      },
      \\"required\\": [
        \\"success\\"
      ]
    },
    \\"GetUserResponse\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        }
      },
      \\"required\\": [
        \\"name\\"
      ]
    }
  }
}"
`;

exports[`JSON Schema generator produces valid code multi-file: yaml 1`] = `
"$schema: 'http://json-schema.org/draft-07/schema#'
definitions:
  CreateUserRequest:
    type: object
    properties:
      name:
        type: string
    required:
      - name
  CreateUserResponse:
    type: object
    properties:
      success:
        type: boolean
    required:
      - success
  GetUserResponse:
    type: object
    properties:
      name:
        type: string
    required:
      - name
"
`;

exports[`JSON Schema generator produces valid code single-file: json 1`] = `
"{
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"definitions\\": {
    \\"CreateUserRequest\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        }
      },
      \\"required\\": [
        \\"name\\"
      ]
    },
    \\"CreateUserResponse\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"success\\": {
          \\"type\\": \\"boolean\\"
        }
      },
      \\"required\\": [
        \\"success\\"
      ]
    },
    \\"GetUserResponse\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\"
        }
      },
      \\"required\\": [
        \\"name\\"
      ]
    }
  }
}"
`;

exports[`JSON Schema generator produces valid code single-file: yaml 1`] = `
"$schema: 'http://json-schema.org/draft-07/schema#'
definitions:
  CreateUserRequest:
    type: object
    properties:
      name:
        type: string
    required:
      - name
  CreateUserResponse:
    type: object
    properties:
      success:
        type: boolean
    required:
      - success
  GetUserResponse:
    type: object
    properties:
      name:
        type: string
    required:
      - name
"
`;
